### PR TITLE
api: Cleanup and improve SubFunction

### LINKED ISSUE #2192: SubFunction. circular dependency
Having a subfunction keep its parent (in `__rkwargs__` and as a property) is a circular dependency problematic for object reconstruction and pickling.

This is partly related to SubFunction handling in general for operations on aprents such as `subs` or `xreplace`

### PR DESCRIPTION
**Pull request by mloubout**
Fixes #2192 

- Fixes `subs/rebuild` of SparseFunction to properly propagate through the subfunctions (coords, ...)
- Add tests

### CODE REVIEW

File: devito/passes/clusters/factorization.py [hunk 1]
```diff
@@ -142,13 +142,16 @@ def run(expr):
...
 
             # Collect common funcs
-            w_funcs = Add(*w_funcs, evaluate=False)
-            w_funcs = collect(w_funcs, funcs, evaluate=False)
-            try:
-                terms.extend([Mul(k, collect_const(v), evaluate=False)
-                              for k, v in w_funcs.items()])
-            except AttributeError:
-                assert w_funcs == 0
+            if len(w_funcs) > 1:
```
Comment (mloubout) [inline devito/passes/clusters/factorization.py:145]: @FabioLuporini  this does not fix the pickle reconstruction issue with `(-1)*` so left the `_print_Mul` in printer for the time being

File: tests/test_interpolation.py [hunk 2]
```diff
@@ -242,8 +242,8 @@ def test_precomputed_injection_time(r):
     sf = PrecomputedSparseTimeFunction(name='s', grid=m.grid, r=r, npoint=len(coords),
                                        gridpoints=gridpoints, nt=nt,
                                        interpolation_coeffs=interpolation_coeffs)
-
-    expr = sf.inject(m, Float(1.))
+    sf.data.fill(1.)
+    expr = sf.inject(m, sf)
```
Comment (mloubout) [inline tests/test_interpolation.py:246]: This one is actually a consequence of removing `parent` as injecting `1` akes an operator that does not process `sf` anymore as an argument as it was only processed through the coordinates as a parent

File: devito/passes/clusters/factorization.py (hunk 1 above)
Comment (FabioLuporini) [reply devito/passes/clusters/factorization.py:145]: OK. Can you open an issue about that and refer to the failing test without the hack?

**Review by FabioLuporini — APPROVED** [review]
Thanks, uncontroversial

**Review by georgebisbas — APPROVED** [review]

**Pull request merged**